dpnp/dpnp_iface_linearalgebra.py

```diff
@@ -269,63 +285,364 @@ def matmul(x1, x2, out=None, **kwargs):
-    if x1_desc and x2_desc and not kwargs:
-        if x1_desc.ndim != 2 or x2_desc.ndim != 2:
-            pass
-        elif not x1_desc.ndim:
-            pass
-        elif not x2_desc.ndim:
-            pass
-        elif not x1_desc.size:
-            pass
-        elif not x2_desc.size:
+    dpnp.check_supported_arrays_type(x1)
+    dpnp.check_supported_arrays_type(x2)
+
+    if subok is False:
+        raise NotImplementedError(
+            "subok keyword argument is only supported by its default value."
+        )
+    else:
+        x1_ndim = x1.ndim
+        x2_ndim = x2.ndim
+
+        if x1_ndim == 0 or x2_ndim == 0:
+            raise ValueError(
+                "matmul: Input operand does not have enough dimensions"
+            )
+
+        exec_q = dpctl.utils.get_execution_queue((x1.sycl_queue, x2.sycl_queue))
+        if exec_q is None:
+            raise ValueError(
+                "Execution placement can not be unambiguously inferred "
+                "from input arguments."
+            )
+
+        # input arrays should be C_CONTIGUOUS or F_CONTIGUOUS
+        if not x1.flags["C_CONTIGUOUS"] and not x1.flags["F_CONTIGUOUS"]:
+            v = dpnp.empty_like(x1, order="C")
+            ht_copy_ev_x1, _ = ti._copy_usm_ndarray_into_usm_ndarray(
+                src=dpnp.get_usm_ndarray(x1),
+                dst=v.get_array(),
+                sycl_queue=x1.sycl_queue,
+            )
+            x1 = v
+            ht_copy_ev_x1.wait()
+
+        if not x2.flags["C_CONTIGUOUS"] and not x2.flags["F_CONTIGUOUS"]:
+            v = dpnp.empty_like(x2, order="C")
+            ht_copy_ev_x2, _ = ti._copy_usm_ndarray_into_usm_ndarray(
+                src=dpnp.get_usm_ndarray(x2),
+                dst=v.get_array(),
+                sycl_queue=x2.sycl_queue,
+            )
+            x2 = v
+            ht_copy_ev_x2.wait()
+
+        squeeze_flag = x1_ndim == 1 or x2_ndim == 1
+        if x1_ndim == 1:
+            x1 = x1[dpnp.newaxis, :]
+            x1_ndim = x1.ndim
+
+        if x2_ndim == 1:
+            x2 = x2[:, dpnp.newaxis]
+            x2_ndim = x2.ndim
+
+        x1_shape = x1.shape
+        x2_shape = x2.shape
+        if x1_shape[-1] != x2_shape[-2]:
+            raise ValueError(
+                "Input operand 1 has a mismatch in its core dimension 0, "
+                "with gufunc signature (n?,k),(k,m?)->(n?,m?) "
+                f"(size {x1_shape[1]} is different from {x2_shape[0]})"
+            )
+
+        # Determine the appropriate data types
+        if dtype is not None:
+            res_dtype = dtype
+            gemm_dtype, _ = _gemm_res_dtype(x1, x2, casting=casting)
+        else:
+            gemm_dtype, res_dtype = _gemm_res_dtype(x1, x2, casting=casting)
+        if x1.dtype != gemm_dtype:
+            x1 = dpnp.astype(x1, gemm_dtype, casting=casting)
+        if x2.dtype != gemm_dtype:
+            x2 = dpnp.astype(x2, gemm_dtype, casting=casting)
```

Comment:
[Tag: copy]
It looks like extra copying
You can set `dtype` during copying before calculating the result
```suggestion
```